### DISCUSSION TITLE
Binding/State performance

### DISCUSSION THREAD

**Discussion by bastiaanterhorst** [Budibase/budibase#4056] [Help]
Hi,
While working on my budibase app, and using more bindings and app state, I'm beginning to notice that these seem to slow the app down quite a bit.

I have two example scenarios I'd like to illustrate:

## Toggling a container based on a state variable
I have a filter block that I want to hide/show. A user clicks a filter icon, container appears. The icon is clicked again, the filtering options disappear.

I do this by conditionally showing the container based on a state variable being true.
I set/invert the state variable as an action on icon click. The action looks like this:
```
if(typeof $("State.show_filter_panel") == 'undefined') {
	return true;
} else {
	if($("State.show_filter_panel") === 'true') {
		return false;
	} else {
		return true;
	}
}
```

While this works very well, it is quite slow! Showing the components takes over a second. Granted, there is a data provider in there that will run, but hiding the component takes just about as long, and the only thing happening is a component being hidden! (I realise I am probably oversimplifying things here). Is saving state a very expensive operation? It seems like that!

## A query that uses multiple bindings

I have a query that responds to filter variables being set. I store these filters in app state. When the user clicks an "apply" button in the filters panel, I set 4 state variables using 4 separate "set value" steps as actions on the button.

The result is that these 4 bindings are set serially. And the result of _that_ is that my query runs 4 time; one time each for every time one of the bindings get changed.

4 query runs might not seem like a lot, but it is still suboptimal. In my specific use case, I loop over records, and then execute a query for each row I get. The state is also used in the nested queries. This means that instead of 12 queries being run, 48 are being run. Quite a lot!

One interesting thing I noticed when looking at the network tab in the dev console, is that it seems to be budibase internals rather than the queries that are making this slow. It takes about 4 seconds after changing filter variables for the queries to run. And when they do, they complete almost instantly, and the UI updates quite quickly.

Am I doing something wrong here? Am I misusing state variables, or bindings?

Cheers,
Bastiaan

**Comment by mjashanks** [comment]
Hi Bastiaan.

### Toggling a container based on a state variable

I have recreated your scenario, but do not see any performance issues. My setup might be much simpler - all I have is
```
> Toggle Button
> Container with condition
   > Data Provider
       > Table, with 10 row limit 
```

It's pretty instant to both show and hide.

What does your structure look like?

It would be great to get an export of your app. It would include data of your internal tables, so I'd understand if that's not possible. You can send it here (if you are able): support@budibase.com

### A query that uses multiple bindings

I agree that 4 state updates should not cause 4 X Query executions - that's a bit mad! Thanks for reporting that.

Again, I cam not seeing a big slow down, even when I set 10 or more state values. Do any of your `Update State` actions include any complex javascript? I have tried using some simple javascript to set all my state values, and it's still instant.

Again - an export of your app would help me debug. Happy to have a call with you as well.

**Reply by bastiaanterhorst** [reply]
Sent you an email, thanks!

**Reply by joebudi** [reply]
Hi @bastiaanterhorst . 
I'm the co-founder of Budibase. Would you be interested in jumping on a quick call to discuss how you are using Budibase, and any feedback you may have? This would mean the world to us, and help us build a better product 👍 .
https://calendly.com/budibase/joe
Cheers,
Joe

**Reply by mjashanks** [reply]
Hey Bastiaan.

I imported your app and setup the DB, and experience the same issue. Theres definitely something funny happening because of 4 x Set states. 

The workaround for now is to replace the 4 Set State actions with 1, that returns a JSON object. I have done this in your app, and it's a lot faster. There is however something wrong with mine - but I haven't gone deeper.

Here is my Set State Javascript, which sets a single JSON object:

```javascript
const filter = {};

filter.show_owner_ids = $("report_settings_form.Fields.owner_ids");
filter.show_date_range = $("report_settings_form.Fields.show_date_range");

var fromdate = new Date();

switch ($("report_settings_form.Fields.show_date_range")) {
  case 'last_7':
		fromdate.setDate(fromdate.getDate() - 7);
		filter.show_date_from = new Date(fromdate.getTime() - (fromdate.getTimezoneOffset() * 60000 )).toISOString().split("T")[0];
    break;
	case 'last_30':
		fromdate.setDate(fromdate.getDate() - 30);
		filter.show_date_from = new Date(fromdate.getTime() - (fromdate.getTimezoneOffset() * 60000 )).toISOString().split("T")[0];
		break;
  case 'last_year':
		fromdate.setFullYear(fromdate.getFullYear() - 1);
		filter.show_date_from = new Date(fromdate.getTime() - (fromdate.getTimezoneOffset() * 60000 )).toISOString().split("T")[0];
		break;
	case 'year_to_date':
		fromdate.setMonth(0,1);
		filter.show_date_from = new Date(fromdate.getTime() - (fromdate.getTimezoneOffset() * 60000 )).toISOString().split("T")[0];
    break;
	// this is the custom clause
  default:
		filter.show_date_from = $("report_settings_form.Fields.date_from");
}


switch ($("report_settings_form.Fields.show_date_range")) {
  case 'last_7':
	case 'last_30':
  case 'last_year':
	case 'year_to_date':
		var todate = new Date();
		filter.show_date_to = new Date(todate.getTime() - (todate.getTimezoneOffset() * 60000 )).toISOString().split("T")[0];
    break;
	// this is the custom clause
  default:
		var todate = new Date($("report_settings_form.Fields.date_to"));
		filter.show_date_to new Date(todate.getTime() - (todate.getTimezoneOffset() * 60000 )).toISOString().split("T")[0];
}

return filter;
``` 

You also have to go and update all the bindings that were previously using the State values. Hopefully this will get you going again :)

**Reply by samg666** [reply]
I met the same performance issue while using a button click to update over 5 app state then navigate to another page, there's noticeable lag about 2s when click the button. It seems app state really an expensive operations, it happens on all of my pages, the lage happens when there more than 3 app state operations on a button actions.

Now, I have stopped using any app state to pass parameters, but replace with URL query string to pass parameters, it much more responsive! Below is my URL query string example, it's much much faster than use app states.  

My button click actions example:

```
let url = "/services/list/"
let pram = new URLSearchParams($("URL.qstring"))
pram.set("Keyword", $("Form1.Fields.SearchKeyword"))
pram.set("Status", $("Form1.Fields.Status"))
pram.set("BillStatus", $("Form1.Fields.BillStatus"))
pram.set("BillCurrency", $("Form1.Fields.BillCurrency"))
pram.set("BillOverdue", $("Form1.Fields.BillOverdue"))
pram.set("OrderBy", btoa($("Form1.Fields.OrderBy")))
pram.set("PageNo", $("Form1.Fields.PageNo"))
pram.set("PageSize", $("Form1.Fields.PageSize"))
return url + encodeURI(pram)
```

Parse query parames:
```
let pram = new URLSearchParams($("URL.qstring"))
return pram.get("ServiceID")
```

**Reply by joebudi** [reply]
Hi @ggs331 
Apologies for interrupting the thread, and thanks for helping out in the community.
I'm the co-founder of Budibase. As I asked Bastiaan above, would you be interested in jumping on a quick call to discuss how you are using Budibase, and any feedback you may have? This would mean the world to us, and help us build a better product 👍 .
https://calendly.com/budibase/joe
Cheers,
Joe

**Comment by bastiaanterhorst** [comment]
Thanks Michael! I have tried your approach, and it works, but unfortunately it does not make my app update any faster.

I have narrowed down the issue to the nested query, and specifically my use of a JS function to display the data returned by the nested query. If I change those to regular handlebar bindings, the delay is gone!

Even when I simply return the data like so:
```
return $("asset_category_stats.getStatsForAssetCategory.current_yield")
```

It is much slower than doing:
```
{{ asset_category_stats.getStatsForAssetCategory.current_yield
```

Is this to be expected? I think I'll try to rewrite my query so I can get all the data I need in a single one. Perhaps that will resolve this issue -- I have not been able to determine that yet.

**Comment by mjashanks** [comment]
FYI Logged this here: https://github.com/Budibase/budibase/issues/4241 - we'll take a look.
